### PR TITLE
[RLlib; Offline RL] - Enable buffering episodes.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Sampling exactly `train_batch_size_per_learner` when using offline data with old API stack `SampleBatch` or new stack `Episode` records is not possible, yet, because each sample batch or episode could contain more than a single timestep. This PR proposes a way to enable sampling with exactly the requested batch size using replay buffers. 

The user can define the replay buffer class to use and its `kwargs`. The `OfflinePreLearner` keeps a replay buffer that buffers episodes and samples from this buffer. The replay buffer serves multiple functions:

1. It buffers multiple episodes from `SampleBatch` or `Episode` data batches and ensures that the requested batch size is sampled.
2. It concatenates different chunks from same episodes and thereby enables the next two points.
3. It enables `n_step` sampling, if needed.
4. It later will enable us to sample data for stateful modules.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm_config.py [hunk 1]
```diff
@@ -2532,17 +2538,45 @@ def offline_data(
...
+                but to evict old experiences as soon as possible.
             prelearner_module_synch_period: The period (number of batches converted)
                 after which the `RLModule` held by the `PreLearner` should sync weights.
                 The `PreLearner` is used to preprocess batches for the learners. The
                 higher this value the more off-policy the `PreLearner`'s module will be.
                 Values too small will force the `PreLearner` to sync more frequently
                 and thus might slow down the data pipeline. The default value chosen
                 by the `OfflinePreLearner` is 10.
-            dataset_num_iters_per_learner: Number of iterations to run in each learner
+            dataset_num_iters_per_learner: Number of updates to run in each learner
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:2576]: 👍

File: rllib/algorithms/dqn/dqn.py [hunk 2]
```diff
@@ -719,6 +719,7 @@ def _training_step_new_api_stack(self, *, with_noise_reset) -> ResultDict:
                         n_step=self.config.n_step,
                         gamma=self.config.gamma,
                         beta=self.config.replay_buffer_config.get("beta"),
+                        sample_episodes=True,
```
Comment (sven1977) [inline rllib/algorithms/dqn/dqn.py:722]: cool!!

File: rllib/offline/tests/test_offline_prelearner.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
+import functools
```
Comment (sven1977) [inline rllib/offline/tests/test_offline_prelearner.py:1]: Wow, thanks for adding all these tests.

**Review by sven1977 — APPROVED** [review]
Awesome PR @simonsays1980 !
Thanks for adding all these tests as well. Offline RL getting stronger by the day.

File: rllib/algorithms/algorithm_config.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/algorithm_config.py:2576]: I admit, `dataset_num_iters_per_learner` is not straight to the point here. `dataset_num_updates_per_learner` is not better in my opinion. Even though `dataset_num_batches_per_learner` describes better that these many different! batches are pulled per learner, but does not point out that it is actually deeply related to an `DataIterator` that iterates these many times. I am not sure, yet. As long as I am not sure, I will leave it as is ;)

**Pull request merged**